api/python/cellxgene_census/src/cellxgene_census/experimental/ml/encoders.py

```diff
@@ -0,0 +1,131 @@
+import abc
+import functools
+from typing import List
+
+import numpy.typing as npt
+import pandas as pd
+from sklearn.preprocessing import LabelEncoder
+
+
+class Encoder(abc.ABC):
+    """Base class for obs encoders.
+
+    To define a custom encoder, five methods must be implemented:
+
+    - ``fit``: defines how the encoder will be fitted to the data.
+    - ``transform``: defines how the encoder will be applied to the data
+      in order to create an obs_tensor.
+    - ``inverse_transform``: defines how to decode the encoded values back
+      to the original values.
+    - ``name``: The name of the encoder. This will be used as the key in the
+      dictionary of encoders. Each encoder passed to a `ExperimentDataPipe` must have a unique name.
+    - ``columns``: List of columns in `obs` that the encoder will be applied to.
+      This will be used to
+
+    See the implementation of ``DefaultEncoder`` for an example.
+    """
+
+    @abc.abstractmethod
+    def fit(self, obs: pd.DataFrame) -> None:
+        """Fit the encoder with obs."""
+        pass
+
+    @abc.abstractmethod
+    def transform(self, df: pd.DataFrame) -> pd.DataFrame:
+        """Transform the obs DataFrame into a DataFrame of encoded values."""
+        pass
+
+    @abc.abstractmethod
+    def inverse_transform(self, encoded_values: npt.ArrayLike) -> npt.ArrayLike:
+        """Inverse transform the encoded values back to the original values."""
+        pass
+
+    @property
+    @abc.abstractmethod
+    def name(self) -> str:
+        """Name of the encoder."""
+        pass
+
+    @property
+    @abc.abstractmethod
+    def columns(self) -> List[str]:
+        """Columns in `obs` that the encoder will be applied to."""
+        pass
+
+
+class DefaultEncoder(Encoder):
```

Comment:
```suggestion
class DefaultEncoder(Encoder):
```

Just a thought, and this is something which is easy to fix with a renaming/ deprecation in the future. Is this a good name?

Why is this called DefaultEncoder? What about it is default?

To me, it's that it's the encoder used by default for columns passed by `obs_column_names`. But I don't think the behavior of a mutually exclusive argument is really the most informative name to give the class. 

Maybe this could be `LabelEncoder` since that name matches the behavior being wrapped?

Reply:
I agree that `LabelEncoder` is probably a better name.